src/server/api/tcp-interface.ts

```diff
@@ -108,11 +90,22 @@ export class BotTunnel {
             this.dataBuffer = undefined;
         }
 
-        const type = parseInt(str.substring(0, 2), 16);
-        const contents = str.substring(3);
+        const packet = packetFromJson(str);
 
-        this.handlePacket(type, contents);
+        if (packet !== null) {
+            // Parse packet based on type
+            switch (packet["type"]) {
+                case "NOTHING": {
+                    break;
+                }
+                case "CLIENT_HELLO": {
+                    this.onHandshake(packet["macAddress"]);
```

Comment:
How statically typed is `packet` here? Are the fields known?

Reply:
TypeScript knows the packet type at this point, see https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions

Reply:
Extremely typed
<img width="306" alt="image" src="https://github.com/Comet-Robotics/chessBot/assets/32407067/0ec59388-8153-4bab-981d-ae59a6d98358">
<img width="377" alt="image" src="https://github.com/Comet-Robotics/chessBot/assets/32407067/d0ec88e1-b8cd-4048-b4cf-acf5bbc04875">

Reply:
Okay - you should probably use `.` notation to access the properties then. IIRC bracket notation has some funky behaviors like allowing you to bypass private in TypeScript (so generally avoided, similar to `==`)